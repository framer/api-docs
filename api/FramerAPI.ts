import {
    RawAPIData,
    AnyModel,
    AnyRawModel,
    Kind,
    RawClassModel,
    RawInterfaceModel,
    RawMethodModel,
    RawEnumModel,
    RawNamespaceModel,
    ClassModel,
    InterfaceModel,
    MethodModel,
    EnumModel,
    NamespaceModel,
    ModelByKind,
} from "./types"

/**
 * Wraps the Framer JSON data and provides methods for accessing the models.
 */
export class FramerAPI {
    constructor(private data: RawAPIData) {}

    /** Look up a model directly by id and return an object with all relationships populated */
    get<T extends AnyModel>(id: string | undefined): T | null {
        if (!id) return null

        const json = (() => {
            const keys = id.toLowerCase().split(".")
            let current = this.data

            while (keys.length) {
                let key = keys.shift()
                if (!key) return null

                // Legacy support for (foo:function) ids that were
                // migrated to (foo:1) in more recent api-extractor
                // versions.
                if (!current[key] && key.endsWith(":function)")) {
                    key = key.replace(":function)", ":1)")
                }
                if (!current[key] && key.endsWith(":instance)")) {
                    key = key.replace(":instance)", ":instance,1)")
                }

                if (!current[key]) return null

                if (keys.length === 0) {
                    return current[key].model
                } else {
                    current = current[key].children
                }
            }

            return null
        })()

        return json ? (this.process(json) as T) : null
    }

    /**
     * Returns all top level models that do not have a parent class or interface.
     */
    rootModels(): ReadonlyArray<AnyModel> {
        return Object.keys(this.data)
            .map(this.get, this)
            .filter(Boolean) as AnyModel[]
    }

    /**
     * Resolves a TSDoc identifier into an APIModel or returns null if
     * unable to find a match. This is useful for both looking up an entry
     * for user input as well as resolving a TSDoc `@link` tag.
     *
     * For example:
     *
     *   // Simple selector
     *   api.resolve('MyClass.someMethod')
     *
     *   // Complex selector
     *   api.resolve('(MyClass:class).(someOverloadedMethod:1)')
     *
     */
    resolve<K extends Kind>(id: string, kind?: K): ModelByKind[K] | null {
        function next(current: RawAPIData, key: string, remainder: string): AnyRawModel | null {
            if (remainder === "") {
                return current[key] ? current[key].model : null
            } else {
                return current[key] ? walk(current[key].children, remainder) : null
            }
        }

        // Walk the tree of raw JSON objects to resolve the path provided, if
        // an ambiguous key has been provided then brute force the lookup by
        // comparing against potential keys.
        function walk(current: RawAPIData, path: string): AnyRawModel | null {
            const [key, ...rest] = path.split(".")
            const remainder = rest.join(".")

            // If the key is wrapped in parens then we assume its a proper id.
            if (key.startsWith("(") && key.endsWith(")")) {
                return next(current, key, remainder)
            } else {
                // If we're on the last item we can use a Kind hint to help
                // differentiate say from a function called Color and a
                // namespace called Color.
                if (kind && remainder === "") {
                    const potentials = keysForKind(key, kind)
                    for (const potential of potentials) {
                        const match = next(current, potential, remainder)
                        if (match) return match
                    }
                    // If provided a hint then bail.
                    return null
                }

                // Otherwise loop through all potential matches
                const potentials = potentialKeys(key)
                for (const potential of potentials) {
                    if (current[potential]) {
                        const match: AnyRawModel | null = next(current, potential, remainder)
                        if (match) return match
                    }
                }
            }

            return null
        }

        const json = walk(this.data, id.toLowerCase())
        return json ? this.process(json) : null
    }

    /**
     * Converts a RawModel into a standard Model, this mostly means replacing
     * id references within the raw representaton with an actual model.
     */
    private process(json: AnyRawModel): AnyModel {
        switch (json.kind) {
            case Kind.Class:
                return this.processClass(json)
            case Kind.Interface:
                return this.processInterface(json)
            case Kind.ConstructSignature:
            case Kind.Constructor:
            case Kind.Function:
            case Kind.Method:
            case Kind.MethodSignature:
                return this.processMethod(json)
            case Kind.Enum:
                return this.processEnum(json)
            case Kind.Namespace:
                return this.processNamespace(json)
            case Kind.EnumMember:
            case Kind.Parameter:
            case Kind.Property:
            case Kind.PropertySignature:
            case Kind.TypeAlias:
            case Kind.Variable:
                return { ...json }
            default:
                return assertNever(json)
        }
    }

    private processClass(json: RawClassModel): ClassModel {
        return {
            ...json,
            constructor: json.constructor ? this.get(json.constructor) : null,
            methods: this.toModels(json.methods),
            properties: this.toModels(json.properties),
        }
    }

    private processInterface(json: RawInterfaceModel): InterfaceModel {
        return {
            ...json,
            constructor: json.constructor ? this.get(json.constructor) : null,
            methods: this.toModels(json.methods),
            properties: this.toModels(json.properties),
        }
    }

    private processMethod(json: RawMethodModel): MethodModel {
        return {
            ...json,
            // Use empty array for overloaded methods otherwise we get recursion here.
            overloads: json.overloadIndex === 0 ? this.toModels(json.overloads) : [],
            parameters: this.toModels(json.parameters),
        }
    }

    private processEnum(json: RawEnumModel): EnumModel {
        return {
            ...json,
            fields: this.toModels(json.fields),
        }
    }

    private processNamespace(json: RawNamespaceModel): NamespaceModel {
        return {
            ...json,
            members: this.toModels(json.members),
        }
    }

    /** Convert an array of model ids to an array of models */
    private toModels<T extends AnyModel[]>(ids: string[]): T {
        return ids.map(this.get, this).filter(Boolean) as any
    }
}

function potentialKeys(name: string): string[] {
    // Strip trailing parentheses if provided, this is a convenience as
    // often people will provide them for methods, e.g. MyClass.method()
    name = name.replace(/\(\)$/, "")
    return [
        `${name}`, // Potential variable, enum field or property
        `(${name}:0)`, // Potential function
        `(${name}:1)`, // Potential function when extending a namespace
        `(${name}:instance)`, // Potential method with no overload
        `(${name}:instance,1)`, // Potential method with overload
        `(${name}:static)`, // Potential static method with no overload
        `(${name}:static,1)`, // Potential static method with overload
        `(${name}:class)`,
        `(${name}:interface)`,
        `(${name}:enum)`,
        `(${name}:function)`,
        `(${name}:variable)`,
        `(${name}:type)`,
        `(${name}:namespace)`,
    ]
}

function keysForKind(name: string, kind: Kind): string[] {
    name = name.replace(/\(\)$/, "")
    switch (kind) {
        case Kind.Class:
            return [`(${name}:class)`]
        case Kind.Interface:
            return [`(${name}:interface)`]
        case Kind.Constructor:
        case Kind.ConstructSignature:
            return [`(${name}:constructor)`]
        case Kind.Function:
            // Support override syntax. We go up to 1 because api-extractor
            // treats functions with namespace extensions as overrides.
            return [`(${name}:function)`, `(${name}:0)`, `(${name}:1)`]
        case Kind.Method:
        case Kind.MethodSignature:
            return [
                `(${name}:instance)`,
                `(${name}:instance,1)`,
                `(${name}:static)`,
                `(${name}:static,1)`,
                `(${name}:0)`,
            ]
        case Kind.Enum:
            return [`(${name}:enum)`]
        case Kind.Namespace:
            return [`(${name}:namespace)`]
        case Kind.EnumMember:
            return [name]
        case Kind.Parameter:
            return [`(${name}:parameter)`]
        case Kind.Property:
            // Support both Property & PropertySignature & MethodSignature here
            return [`(${name}:instance)`, `(${name}:0)`, name]
        case Kind.PropertySignature:
            return [name]
        case Kind.TypeAlias:
            return [name]
        case Kind.Variable:
            return [`(${name}:variable)`, name]
        default:
            return assertNever(kind)
    }
}

function assertNever(x: never): never {
    throw new Error("Unexpected object: " + x)
}
