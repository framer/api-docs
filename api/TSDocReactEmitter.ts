import * as React from "react"
import { renderToStaticMarkup } from "react-dom/server"
import {
    DocNode,
    DocNodeKind,
    DocBlock,
    DocBlockTag,
    DocCodeSpan,
    DocComment,
    DocDeclarationReference,
    DocErrorText,
    DocEscapedText,
    DocFencedCode,
    DocHtmlStartTag,
    DocInlineTag,
    DocLinkTag,
    DocSection,
    DocParagraph,
    DocNodeTransforms,
    DocParamBlock,
    DocParamCollection,
    DocPlainText,
    PlainTextEmitter,
    TSDocEmitter,
    StringBuilder,
} from "@microsoft/tsdoc"

type CreateElementFn<T> = (tag: string | object, props?: object, children?: T | T[], ...rest: T[]) => T
type FragmentId = string | object
type LikeReact<T> = { createElement: CreateElementFn<T>; Fragment: FragmentId }

// Converts a TSDoc DocNode into an HTML string
// http://api-extractor.com/pages/tsdoc/syntax/
// https://github.com/Microsoft/tsdoc/blob/6034bee/spec/code-snippets/DeclarationReferences.ts
// https://github.com/Microsoft/tsdoc/tree/6034bee/tsdoc/src/emitters
// https://microsoft.github.io/tsdoc/#
export function renderTSDocToHTML(node: DocNode | undefined): string {
    const element = render(node, React)
    if (!element) return ""
    if (!React.isValidElement(element)) throw new Error(`render() returned non-ReactElement: ${typeof element}`)
    return renderToStaticMarkup(element)
}

// Converts a TSDoc DocNode into a tree, usually React but allows custom createElement and Fragment methods.
export function renderTSDoc<T>(node: DocNode | undefined, opts: LikeReact<T>) {
    return render(node, opts)
}

function render<T>(docNode: DocNode | undefined, { createElement, Fragment }: LikeReact<any>): T | null {
    function renderNode(node: DocNode | undefined): T | null {
        return render(node, { createElement, Fragment })
    }
    function renderNodes(docNodes: ReadonlyArray<DocNode | undefined>): T[] {
        const rendered: (T | null)[] = []

        let idx = 0
        let n: DocNode | undefined

        // TODO: Re-write this to not be terrible.
        outer: while ((n = docNodes[idx])) {
            // TSDoc may include HTML markup, it processes this in a flat stream of DocNodes
            // that may look like [HtmlStartTag, HtmlStartTag, PlainText, HtmlEndTag, HtmlEndTag]
            // which is fine if you're rendering everything in series but we need to handle the
            // nesting manually for React.createElement(). So we break out an inner loop here
            // to process the children.
            if (n.kind === DocNodeKind.HtmlStartTag) {
                const html = n as DocHtmlStartTag

                const attrs: { [attr: string]: string } = {}
                for (const attr of html.htmlAttributes) {
                    attrs[attr.name] = attr.value
                }

                // Bail early if we find a self-closing tag.
                if (html.selfClosingTag) {
                    rendered.push(createElement(html.name, attrs))
                    idx += 1
                    continue
                }

                // Collect the children.
                let child: DocNode | undefined
                let childIdx = idx + 1 // Move on to next node and process children until we find a closing tag.
                let depth = 0 // Track the depth of nested HtmlStartTags
                const children: DocNode[] = []

                while (!html.selfClosingTag && (child = docNodes[childIdx])) {
                    if (child.kind === DocNodeKind.HtmlStartTag && !(child as DocHtmlStartTag).selfClosingTag) {
                        depth += 1
                    } else if (child.kind === DocNodeKind.HtmlEndTag) {
                        if (depth === 0) {
                            idx = childIdx + 1 // update parent index and continue the outer loop
                            rendered.push(createElement(html.name, attrs, renderNodes(children)))
                            continue outer
                        } else {
                            depth -= 1
                        }
                    }
                    children.push(child)
                    childIdx += 1
                }

                rendered.push(createElement(html.name, attrs))
            } else {
                rendered.push(renderNode(n))
            }
            idx += 1
        }

        // Need to provide a key for each element...
        return rendered.map((child, idx) => createElement(Fragment, { key: idx }, child))
    }

    if (docNode === undefined) {
        return null
    }

    // NOTE: Fragments are used to keep the HTML structure as flat as possible for styling.
    // ideally we want just one level of block elements, e.g <p>, <pre> etc.
    // We use the `key` attribute to tag the fragments to help identify the type of
    // DocNodeKind that created it for debugging.
    const kind = docNode.kind as DocNodeKind

    switch (kind) {
        // A semantic group of content denoted by a block tag
        case DocNodeKind.Block:
            const docBlock: DocBlock = docNode as DocBlock

            return createElement(Fragment, { key: "DocBlock" }, renderNode(docBlock.content))

        // A block tag, e.g. @remarks or @example
        case DocNodeKind.BlockTag:
            return null // We don't want to show @ tags

        // Inline code
        case DocNodeKind.CodeSpan:
            const docCodeSpan: DocCodeSpan = docNode as DocCodeSpan
            return createElement("code", { className: `DocCodeSpan` }, docCodeSpan.code)

        // Full TSDoc Comment denoted by /**  */
        case DocNodeKind.Comment:
            const docComment: DocComment = docNode as DocComment
            const content = renderNodes([
                ...docComment.customBlocks,
                docComment.summarySection,
                docComment.remarksBlock,
                docComment.privateRemarks,
                docComment.deprecatedBlock,
                docComment.params,
                docComment.typeParams,
                docComment.returnsBlock,
                docComment.inheritDocTag,
            ])
            if (docComment.modifierTagSet.nodes.length > 0) {
                //this._ensureLineSkipped()
                content.push(...renderNodes(docComment.modifierTagSet.nodes))
            }
            return createElement(Fragment, { key: "DocComment" }, content)

        // Package part of a link reference e.g. "pkg" in {@link pkg#MyClass.Button}
        // http://api-extractor.com/pages/tsdoc/syntax/#api-item-references
        case DocNodeKind.DeclarationReference:
            const docDeclarationReference: DocDeclarationReference = docNode as DocDeclarationReference
            let reference = `${docDeclarationReference.packageName || ""}${docDeclarationReference.importPath || ""}`
            if (docDeclarationReference.packageName !== undefined || docDeclarationReference.importPath !== undefined) {
                reference += "#"
            }
            return createElement(
                "code",
                { className: "DocDeclarationReference" },
                createElement("a", { "data-link-ref": "" }, [
                    reference,
                    ...renderNodes(docDeclarationReference.memberReferences),
                ])
            )

        // Error text...
        case DocNodeKind.ErrorText:
            const docErrorText: DocErrorText = docNode as DocErrorText
            return createElement("span", { className: "DocErrorText" }, docErrorText.text)

        // ???
        case DocNodeKind.EscapedText:
            const docEscapedText: DocEscapedText = docNode as DocEscapedText
            return createElement(Fragment, { key: "DocEscapedText" }, docEscapedText.encodedText)

        // Block level code denoted by three back-ticks ```tsx
        case DocNodeKind.FencedCode:
            const docFencedCode: DocFencedCode = docNode as DocFencedCode
            return createElement(
                "pre",
                { className: "DocFencedCode" },
                createElement(
                    "code",
                    {
                        "data-lang": docFencedCode.language,
                        className: `language-${docFencedCode.language}`,
                    },
                    docFencedCode.code
                )
            )

        case DocNodeKind.HtmlAttribute:
            throw new Error("Encountered unexpected HtmlAttribute. Should be handled by renderNodes()")

        case DocNodeKind.HtmlEndTag:
            throw new Error("Encountered unexpected HtmlEndTag. Should be handled by renderNodes()")

        case DocNodeKind.HtmlStartTag:
            throw new Error("Encountered unexpected HtmlStartTag. Should be handled by renderNodes()")

        // @inherit
        case DocNodeKind.InheritDocTag:
            //const docInheritDocTag: DocInheritDocTag = docNode as DocInheritDocTag
            return null // We don't want to render this tag

        // Inline custom tag e.g @foo can be wrapped in curly braces to include content e.g. {@foo bar}
        // http://api-extractor.com/pages/tsdoc/syntax/#inline-tags
        case DocNodeKind.InlineTag:
            const docInlineTag: DocInlineTag = docNode as DocInlineTag
            let docInlineContent = ""
            if (docInlineTag.tagContent.length > 0) {
                docInlineContent = " " + docInlineTag.tagContent
            }
            return createElement(Fragment, { key: "DocInlineTag" }, `{${docInlineTag.tagName}${docInlineContent}}`)

        // Inline @link tag.
        case DocNodeKind.LinkTag:
            const docLinkTag: DocLinkTag = docNode as DocLinkTag
            if (docLinkTag.codeDestination) {
                const ref = declarationReference(docLinkTag.codeDestination)
                const name = createElement("code", {}, declarationReferenceDisplayName(docLinkTag.codeDestination))
                return createElement("a", { className: "DocRefTag", "data-link-ref": ref }, docLinkTag.linkText || name)
            }
            let href = docLinkTag.urlDestination || "#"
            return createElement("a", { className: "DocLinkTag", href }, docLinkTag.linkText)

        // The "MyClass" and "foo" part of a {@link MyClass.Foo} reference
        case DocNodeKind.MemberIdentifier:
            throw new Error("Should be handled by renderDeclarationReference")

        // The "MyClass.Foo" part of a {@link MyClass.Foo} reference
        case DocNodeKind.MemberReference:
            throw new Error("Should be handled by renderDeclarationReference")

        // Part of a selectior {@link MyClass.(foo:1)} reference
        case DocNodeKind.MemberSelector:
            throw new Error("Should be handled by renderDeclarationReference")

        case DocNodeKind.MemberSymbol:
            throw new Error("Should be handled by renderDeclarationReference")

        // Every DocBlock has a corresponding section.
        case DocNodeKind.Section:
            const docSection: DocSection = docNode as DocSection
            if (docSection.nodes.length === 0 || !PlainTextEmitter.hasAnyTextContent(docSection)) return null
            return createElement(Fragment, { key: "DocSection" }, renderNodes(docSection.nodes))

        // A chunk of text seperated by two newlines.
        case DocNodeKind.Paragraph:
            const listDelimiter = /[*-]\s*/
            const trimmedParagraph: DocParagraph = DocNodeTransforms.trimSpacesInParagraph(docNode as DocParagraph)
            if (trimmedParagraph.nodes.length === 0) return null

            // TSDoc doesn't support lists but we'd like to render markdown style ones so if a paragraph
            // begins with a dash we break it down into list items by splitting on hyphens or asterisks.
            const firstChild = trimmedParagraph.nodes[0]
            const firstChildText = firstChild.kind === DocNodeKind.PlainText ? (firstChild as DocPlainText).text : null
            if (!firstChildText || firstChildText.search(listDelimiter) !== 0) {
                return createElement("p", {}, renderNodes(trimmedParagraph.nodes))
            }

            const items: DocNode[][] = []
            let fragments: DocNode[] = []
            trimmedParagraph.nodes.forEach(child => {
                if (child.kind === DocNodeKind.PlainText) {
                    const textNode = child as DocPlainText
                    if (listDelimiter.test(textNode.text)) {
                        const listText = textNode.text.split(listDelimiter)

                        // Create a list item with first part of the current
                        // text string plus previous fragments.
                        if (fragments.length || listText[0]) {
                            const textNode = new DocPlainText({
                                configuration: docNode.configuration,
                                text: listText.shift() || "",
                            })
                            items.push([...fragments, textNode])
                            fragments = []
                        }

                        // Save the last fragment for later
                        if (listText.length) {
                            const textNode = new DocPlainText({
                                configuration: docNode.configuration,
                                text: listText.pop() || "",
                            })

                            fragments.push(textNode)
                        }

                        // Create list items for each of the middle text strings.
                        const listItems = listText
                            .filter(text => text.trim())
                            .map(text => [new DocPlainText({ configuration: docNode.configuration, text })])
                        items.push(...listItems)
                    } else {
                        fragments.push(child)
                    }
                } else {
                    fragments.push(child)
                }
            })

            // Append remaining fragments
            if (fragments.length) {
                // If we have only fragments add a new line entry.
                if (items.length === 0) {
                    items.push([])
                }

                // If the first fragment is an empty string it's a new line
                // marker. Append a new list entry.
                if (fragments[0].kind === DocNodeKind.PlainText) {
                    const textNode = fragments[0] as DocPlainText
                    if (textNode.text === "") {
                        items.push([])
                    }
                }
                items[items.length - 1].push(...fragments)
            }

            const children = items.map(items => createElement("li", { key: id() }, renderNodes(items)))
            return createElement("ul", {}, children)

        // Defined by a @param
        // http://api-extractor.com/pages/tsdoc/syntax/#param
        case DocNodeKind.ParamBlock:
            const docParamBlock: DocParamBlock = docNode as DocParamBlock
            return createElement(
                "div",
                { className: "DocParamBlock" },
                createElement(Fragment, {}, docParamBlock.parameterName),
                " ",
                renderNode(docParamBlock.content)
            )

        // A collection of @param tags
        case DocNodeKind.ParamCollection:
            const docParamCollection: DocParamCollection = docNode as DocParamCollection
            return createElement(Fragment, { key: "DocParamCollection" }, renderNodes(docParamCollection.blocks))

        // Just good old fashioned text.
        case DocNodeKind.PlainText:
            const docPlainText: DocPlainText = docNode as DocPlainText
            return createElement(Fragment, { key: "DocPlainText" }, docPlainText.text)
        case DocNodeKind.Excerpt:
            return null
        // A single newline
        case DocNodeKind.SoftBreak:
            return null
        default:
            return assertNever(kind)
    }
}

function assertNever(x: never): never {
    throw new Error("Unexpected object: " + x)
}

let moduleId = 0
function id() {
    return ++moduleId
}

function declarationReferenceDisplayName(node: DocDeclarationReference): string {
    let name: string[] = []
    for (const member of node.memberReferences) {
        if (member.hasDot) {
            name.push(".")
        }
        if (member.memberIdentifier) {
            name.push(member.memberIdentifier.identifier)
        }
    }
    return name.join("")
}

function declarationReference(node: DocDeclarationReference) {
    const emitter = new TSDocEmitter()
    const builder = new StringBuilder()
    emitter.renderDeclarationReference(builder, node)
    return builder.toString()
}
