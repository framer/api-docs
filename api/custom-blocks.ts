import { AedocDefinitions } from "@microsoft/api-extractor-model"
import { TSDocConfiguration, TSDocTagDefinition, TSDocTagSyntaxKind } from "@microsoft/tsdoc"

/**
 * Add support for `@production` and `@prototype` documentation blocks.
 *
 * We monkey-patch `AedocDefinitions` which `APIModel` uses to configure its block support.
 * There currently isn't an external API to add this behaviour.
 */

export const configuration: TSDocConfiguration = AedocDefinitions.tsdocConfiguration

const production = new TSDocTagDefinition({
    tagName: "@production",
    syntaxKind: TSDocTagSyntaxKind.BlockTag,
    allowMultiple: false,
})

const prototype = new TSDocTagDefinition({
    tagName: "@prototype",
    syntaxKind: TSDocTagSyntaxKind.BlockTag,
    allowMultiple: false,
})

configuration.addTagDefinitions([production, prototype], true)

configuration.setSupportForTags([production, prototype], true)

export function patchAPIExtractorWithCustomTSDocTags() {
    Object.defineProperty(AedocDefinitions, "tsdocConfiguration", {
        get: () => configuration,
    })
}
