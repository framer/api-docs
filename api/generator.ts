// Helper script to generate the framer.data.json file that is then consumed by
// the FramerAPIContext component. The JSON file should extract as much useful
// information from the api-extractor AST and it's nested TSDoc AST to avoid
// having to do all this processing in the React components.
//
// The script takes a list of inputs - api.json files generated by api-extractor
// and will output a processed JSON file to stdout.
//
// Usage:
//   % ./api/generator.ts <input> [...<input2>]
//
// References:
// https://github.com/Microsoft/web-build-tools/blob/master/apps/api-documenter/src/documenters/MarkdownDocumenter.ts
// https://microsoft.github.io/web-build-tools/api/api-extractor.html
//
// To Do:
// - [ ] Support an optional --out flag to specify an output file.
import * as fs from "fs"
import * as path from "path"
import * as assert from "assert"
import { ApiModel, ReleaseTag as ApiReleaseTag } from "@microsoft/api-extractor-model"
import { walk } from "./walker"
import { RawAPIData, Kind, AnyRawModel } from "./types"

function usage() {
    console.log(`Usage:\n    ${path.basename(process.argv[1])} <input1>[, <input2>]\n`)
    console.log(
        `Example:\n    ${path.basename(
            process.argv[1]
        )} ../FramerStudio/Source/Library/build/framer.api.json api/framer.data.json\n`
    )
}

if (process.argv.length <= 2) {
    usage()
    process.exit()
}

function processInput(filepath: string): RawAPIData {
    if (!fs.existsSync(filepath)) {
        console.error(`Unable to read input file: ${filepath}`)
        process.exit(1)
    }

    const pkg = new ApiModel().loadPackage(filepath)
    assert.equal(pkg.entryPoints.length, 1, "Only support a single entry point at the moment")

    const map: RawAPIData = {}
    walk(pkg.entryPoints[0], ApiReleaseTag.Beta, item => {
        type PartialRawAPIData = { [key: string]: { model?: AnyRawModel; children: PartialRawAPIData } }
        const keys = item.id.split(".")

        // Holds the current map of key -> child
        let current: PartialRawAPIData = map
        while (keys.length) {
            const key = keys.shift()!

            if (!current[key]) {
                current[key] = { children: {} }
            }

            if (keys.length === 0) {
                // Warn if we encounter duplicate fields exported by seperate packages
                // we can ignore params though as these will regularly crop up due to
                // overloads.
                // TODO: We could include the package name with the namespace.
                if (current[key].model && item.kind !== Kind.Parameter) {
                    process.stderr.write(`Skipping duplicate entry for "${item.id}"\n`)
                } else {
                    current[key].model = item
                }
            }

            current = current[key].children
        }
    })

    return map
}

let map: RawAPIData = {}
for (const filepath of process.argv.slice(2)) {
    map = { ...map, ...processInput(filepath) }
}

const data = JSON.stringify(map, null, "  ") + "\n"
process.stdout.write(data)
