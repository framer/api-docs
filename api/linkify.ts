// Processes the built site to wire up any @link tokens included in the TSDoc
// source code. It does this in two passes, the first to build up a page
// of paths -> api item id, the second to append an href to all @link tags
// the modified document is then written over the original file.
//
// For the other implementation points see the `LinkTag` geneation in
// TSDocReactEmitter.ts and the id resolution in the APIOverviewElement
// component.
//
// Usage:
//
//     % yarn ts-node ./linkify.ts <htmlfile> [, <htmlfile>...]
//     % yarn ts-node ./linkify.ts build/**/*.html
import * as fs from "fs"
import * as cheerio from "cheerio"
import chalk from "chalk"

const { yellow, gray, white } = chalk

const inputs = process.argv.slice(2)

const parsed = inputs.map(filepath => cheerio.load(fs.readFileSync(filepath).toString("utf-8")))
const pathById: { [id: string]: string } = {}

console.log("Updating TSDoc @link references")

parsed.forEach($ => {
    $("[data-permalink-ref][data-permalink-path]").each((_, el) => {
        const { permalinkRef, permalinkPath } = $(el).data()
        pathById[permalinkRef] = permalinkPath
    })
})

parsed.forEach(($, idx) => {
    const filepath = inputs[idx]
    let hasMatch = false

    $("[data-link-ref]").each((_, el) => {
        hasMatch = true

        const $el = $(el)
        const ref = $el.data("linkRef")
        if (pathById[ref]) {
            $el.attr("href", `${pathById[ref]}`)
        } else {
            $el.addClass("link-ref-missing")
            const id = $el.parents("[data-tsdoc-ref]").data("tsdocRef")
            console.warn(
                gray(
                    `${yellow("Warning:")} Could not resolve reference for: ${white(ref)} in ${white(
                        id || "Unknown"
                    )} on page ${white(filepath)}`
                )
            )
        }
    })

    if (hasMatch) {
        console.log(gray(`Writing new @link references â†’ ${white(filepath)}`))
        fs.writeFileSync(filepath, $.html())
    }
})
