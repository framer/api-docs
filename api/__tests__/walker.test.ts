import * as json from "../__fixtures__/example.api.json"
import { ApiPackage, ApiItem, ReleaseTag } from "@microsoft/api-extractor-model"
import { walk } from "../walker"

describe("walk", () => {
    function createApiPackage() {
        return ApiItem.deserialize(json as any) as ApiPackage
    }

    it("should generate ids", () => {
        const ids: string[] = []
        const pkg = createApiPackage()
        walk(pkg.entryPoints[0], ReleaseTag.Public, item => ids.push(item.id))
        expect(ids).toMatchSnapshot()
    })
})
