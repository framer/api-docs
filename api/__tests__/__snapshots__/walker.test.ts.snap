// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`walk should generate ids 1`] = `
Array [
  "(myclass:class)",
  "(myclass:constructor)",
  "(myclass:constructor).(param1:parameter)",
  "(myclass:class).(mymethod:instance,1)",
  "(myclass:class).(mymethod:instance,1).(param1:parameter)",
  "(myclass:class).(mymethod:instance,2)",
  "(myclass:class).(mymethod:instance,2).(overload1:parameter)",
  "(myclass:class).(myproperty:instance)",
  "(myclass:class).(mystaticmethod:static,1)",
  "(myclass:class).(mystaticmethod:static,1).(param1:parameter)",
  "(myclass:class).(mystaticmethod:static,2)",
  "(myclass:class).(mystaticmethod:static,2).(overloaded1:parameter)",
  "(myclass:class).(mystaticproperty:static)",
  "(myenum:enum)",
  "(myenum:enum).myfield1",
  "(myenum:enum).myfield2",
  "(myenum:enum).myfield3",
  "(myextendedenum:enum)",
  "(myextendedenum:enum).myextendedenumfield",
  "(myextendedenum:namespace)",
  "(myextendedenum:namespace).(myextendedenummethod:function)",
  "(myfunction:1)",
  "(myfunction:1).(param1:parameter)",
  "(myfunction:2)",
  "(myfunction:2).(overloaded1:parameter)",
  "(myfunctionwithnamespace:1)",
  "(myfunctionwithnamespace:1).(param1:parameter)",
  "(myfunctionwithnamespace:namespace)",
  "(myfunctionwithnamespace:namespace).(myfunctionwithnamespacefunction:function)",
  "(myfunctionwithnamespace:namespace).(myfunctionwithnamespacefunction:function).(param1:parameter)",
  "(myinterface:interface)",
  "(myinterface:interface).(:new,0)",
  "(myinterface:interface).(:new,0).(param1:parameter)",
  "(myinterface:interface).(mymethodsignature:0)",
  "(myinterface:interface).(mymethodsignature:0).(param1:parameter)",
  "(myinterface:interface).mypropertysignature",
  "(mynamespace:namespace)",
  "(mynamespace:namespace).(mynamespacefunction:function)",
  "(mynamespace:namespace).(mynamespacefunction:function).(param1:parameter)",
  "(mynamespace:namespace).(mynamespaceinterface:interface)",
  "(mynamespace:namespace).mynamespacetype",
  "(mynamespace:namespace).(mynamespacevariable:variable)",
  "(my_variable:variable)",
  "mytype",
]
`;
